File: advent_of_code/day2/part1/main.py
class Policy(object):

    def __init__(
        self, 
        lower_limit: int = 0, 
        upper_limit: int = 0, 
        character: str = "", 
        password: str = ""
    ) -> None:
        self.lower_limit = lower_limit
        self.upper_limit = upper_limit
        self.character = character
        self.password = password

    def is_password_valid(self) -> bool:
        return self.lower_limit <= self.password.count(self.character) <= self.upper_limit



def make_policy(line: str) -> Policy:

    p = Policy()
    split_line = line.split()

    p.lower_limit = int(split_line[0].split("-")[0])
    p.upper_limit = int(split_line[0].split("-")[1])
    p.character = split_line[1][0]
    p.password = split_line[-1]

    return p

def main(file: str):
    pass

File: advent_of_code/day2/part1/main.go
package part1

import (
	"advent-of-code/advent_of_code/helpers"
	"fmt"
	"strconv"
	"strings"
)

type Policy struct {
	LowerLimit int
	UpperLimit int
	Character  string
	Password   string
	IsValid    bool
}

//  tbd
func makePolicy(line string) *Policy {
	// 1-3 a: abcde
	var policy Policy = Policy{}

	splitLine := strings.Fields(line)
	a := strings.Split(splitLine[0], "-")[0]
	b := strings.Split(splitLine[0], "-")[1]
	c := strings.Split(splitLine[1], ":")[0]
	d := splitLine[2]

	aa, _ := strconv.ParseInt(a, 10, 64)
	bb, _ := strconv.ParseInt(b, 10, 64)

	policy.LowerLimit = int(aa)
	policy.UpperLimit = int(bb)
	policy.Character = c
	policy.Password = d

	return &policy
}

func isPolicyValid(p Policy) bool {
	if p.LowerLimit <= strings.Count(p.Password, p.Character) && strings.Count(p.Password, p.Character) <= p.UpperLimit {
		p.IsValid = true
		return true
	}
	return false
}

func main(filepath string) int {
	lines, err := helpers.ReadFileLineByLine(filepath)
	if err != nil {
		fmt.Println(err)
	}

	var results int

	for _, line := range lines {
		policy := makePolicy(line)
		if isPolicyValid(*policy) {
			results++
		}
	}

	return results

}
